File: 0x10 基本数据结构/0x13 链表与邻接表/AcWing137.py
N = 100006
n, tot, P = 0, 0, 99991
a, snow, head, nxt = [0]*6, [0]*N, [0]*N, [0]*N
def H(a):
    global n, tot, P
    sum_, mul_ = 0, 1
    for i in range(6):
        sum_ = (sum_ + a[i]) % P
        mul_ = mul_ * a[i] % P
    return (sum_ + mul_) % P
    
def equal(a, b):
    global n, tot, P
    for i in range(6):
        for j in range(6):
            eq = 1
            for k in range(6):
                if a[(i+k)%6] != b[(j+k)%6]:
                    eq = 0
                    break
            if eq: return 1
            eq = 1
            for k in range(6):
                if a[(i+k)%6] != b[(j-k+6)%6]:
                    eq = 0
                    break
            if eq: return 1
    return 0
    
def insert(a):
    global n, tot, P
    val = H(a)
    i = head[val]
    while i:
        if equal(snow[i], a): return 1
        i = nxt[i]
    tot += 1
    snow[tot] = a[:]
    nxt[tot] = head[val]
    head[val] = tot
    return 0
    
def main():
    global n, tot, P
    n = int(input())
    for i in range(1, n+1):
        a[:6] = [int(x) for x in input().split()]
        if insert(a):
            print("Twin snowflakes found.")
            return
    print("No two snowflakes are alike.")
    
main()

File: 0x10 基本数据结构/0x14 Hash/AcWing137.cpp
#include <cstring>
#include <iostream>
#define ll long long
using namespace std;
const int N = 100006;
int n, tot, P = 99991; // 随机数据期望O(N^2/P)，取P为接近的大质数
int snow[N][6], head[N], nxt[N];
int H(int *a) {
  int sum = 0, mul = 1;
  for (int i=0; i<6; i++) {
    sum = (sum + a[i]) % P;
    mul = (ll)mul * a[i] % P;
  }
  return (sum + mul) % P;
}

bool equal(int *a, int *b) {
  for (int i=0; i<6; i++)
    for (int j=0; j<6; j++) {
      bool eq = 1;
      for (int k=0; k<6; k++)
        if (a[(i+k)%6] != b[(j+k)%6]) {
          eq = 0; // 顺时针对比
          break;
        }
      if (eq) return 1;
      eq = 1;
      for (int k=0; k<6; k++)
        if (a[(i+k)%6] != b[(j-k+6)%6]) {
          eq = 0; // 逆时针对比
          break;
        }
      if (eq) return 1;
    }
  return 0;
}

bool insert(int *a) {
  int val = H(a);
  // 遍历表头head[val]指向的链表，在同一slot中寻找形状相同的雪花
  for (int i=head[val]; i; i=nxt[i])
    if (equal(snow[i], a)) return 1;
  // 未找到形状相同的雪花，执行插入
  ++tot;
  memcpy(snow[tot], a, 6*sizeof(int));
  nxt[tot] = head[val], head[val] = tot;
  return 0;
}

int main() {
  cin >> n;
  for (int i=1; i<=n; i++) {
    int a[10];
    for (int j=0; j<6; j++) scanf("%d", &a[j]);
    if (insert(a)) {
      puts("Twin snowflakes found.");
      return 0;
    }
  }
  puts("No two snowflakes are alike.");
  return 0;
}